### PR TITLE
[MLIR] Improve KernelOutlining to avoid introducing an extra block

### PR DESCRIPTION
**Pull request by joker-eph**
This fixes a TODO in the code.

### CODE REVIEW

File: mlir/lib/Dialect/GPU/Transforms/KernelOutlining.cpp [hunk 1]
```diff
@@ -241,24 +241,19 @@ static gpu::GPUFuncOp outlineKernelFuncImpl(gpu::LaunchOp launchOp,
...
-    OpBuilder replacer(op);
-    replacer.create<gpu::ReturnOp>(op.getLoc());
-    op.erase();
-  });
+  // Splice now the entry block of the gpu.launch operation at the end of the
+  // gpu.func entry block and erase the redundant block.
+  Block *clonedLaunchOpEntry = map.lookup(&launchOpBody.front());
+  Operation *terminator = clonedLaunchOpEntry->getTerminator();
+  OpBuilder replacer(terminator);
+  replacer.create<gpu::ReturnOp>(terminator->getLoc());
```
Comment (fabianmcg) [inline mlir/lib/Dialect/GPU/Transforms/KernelOutlining.cpp:251]: `LaunchOp::body`  is marked as `AnyRegion`, isn't this assuming a single block? It'd be in favor of changing it into having a single block.

Comment (fabianmcg) [reply mlir/lib/Dialect/GPU/Transforms/KernelOutlining.cpp:251]: Coincidentally we don't have any outlining tests with multiple blocks.

Comment (joker-eph) [reply mlir/lib/Dialect/GPU/Transforms/KernelOutlining.cpp]: Added a test. I am not sure why we should just arbitrarily restrict it? The body of the gpu.launch is like a function body to me.

Comment (fabianmcg) [reply mlir/lib/Dialect/GPU/Transforms/KernelOutlining.cpp]: I've always seen `launch` as a high-level operation with structured control flow, however, I guess most code-bases assume the function body so I agree there's no reason for restricting it.

**Review by fabianmcg — APPROVED** [review]
LGTM.

fyi, a test is failing.

**Review by grypp — APPROVED** [review]
Nice, thanks for improving that

**Pull request merged**